### ISSUE TITLE
JaySvcUtil 1.0 RC1 question

### ISSUE THREAD

**Issue by RainerAtSpirit** [jaystack/jaydata#9]
Hi there,

I'm following along your OData demo at http://jaydata.org/tutorials/how-to-build-a-simple-odata-based-ajax-application, but I would like do use existing services instead of a V11 one. 
1. http://odata.netflix.com/v2/catalog/$metadata a well known OData V2.0 service ;-)
2. http://openaccess.spirit.de/EntitiesModelService.svc/$metadata an OData service generated by OpenAccess

When running the tool for 1 I'm getting the following warnings, not sure if these are expected.

```
Requesting: http://odata.netflix.com/v2/catalog/$metadata... done.
Info: generating type Netflix.Catalog.v2.Genre
  Warning: Genre.Name:m:FC_TargetPath is an unknown/unprocessed attribued
  Warning: Genre.Name:m:FC_ContentKind is an unknown/unprocessed attribued
  Warning: Genre.Name:m:FC_KeepInContent is an unknown/unprocessed attribued
Info: generating type Netflix.Catalog.v2.Title
  Warning: Title.Name:m:FC_TargetPath is an unknown/unprocessed attribued
  Warning: Title.Name:m:FC_ContentKind is an unknown/unprocessed attribued
  Warning: Title.Name:m:FC_KeepInContent is an unknown/unprocessed attribued
  Warning: Title.Synopsis:m:FC_TargetPath is an unknown/unprocessed attribued
  Warning: Title.Synopsis:m:FC_ContentKind is an unknown/unprocessed attribued
  Warning: Title.Synopsis:m:FC_KeepInContent is an unknown/unprocessed attribued
  Warning: Title.DateModified:m:FC_TargetPath is an unknown/unprocessed attribued
  Warning: Title.DateModified:m:FC_ContentKind is an unknown/unprocessed attribued
  Warning: Title.DateModified:m:FC_KeepInContent is an unknown/unprocessed attribued
  Warning: inverseProperty other side missing: Netflix.Catalog.v2.Title_Disc
  Warning: inverseProperty other side missing: Netflix.Catalog.v2.Title_Season
  Warning: inverseProperty other side missing: Netflix.Catalog.v2.Title_Series
  Warning: inverseProperty other side missing: Netflix.Catalog.v2.Title_Movie
Info: generating type Netflix.Catalog.v2.DeliveryFormatAvailability
Info: generating type Netflix.Catalog.v2.BoxArt
Info: generating type Netflix.Catalog.v2.InstantAvailability
Info: generating type Netflix.Catalog.v2.TitleAudioFormat
Info: generating type Netflix.Catalog.v2.TitleAward
Info: generating type Netflix.Catalog.v2.Person
  Warning: Person.Name:m:FC_TargetPath is an unknown/unprocessed attribued
  Warning: Person.Name:m:FC_ContentKind is an unknown/unprocessed attribued
  Warning: Person.Name:m:FC_KeepInContent is an unknown/unprocessed attribued
Info: generating type Netflix.Catalog.v2.TitleScreenFormat
Info: generating type Netflix.Catalog.v2.Language
  Warning: Language.Name:m:FC_TargetPath is an unknown/unprocessed attribued
  Warning: Language.Name:m:FC_ContentKind is an unknown/unprocessed attribued
  Warning: Language.Name:m:FC_KeepInContent is an unknown/unprocessed attribued
```

When running the tool for 2 all looks good, but the in the generated JavaScript file the entity information for `Person`and `Company`is missing.

``` javascript

////////////////////////////////////////////////////////////////////////////////////////
////// Autogenerated - do not modify or be prepared
////////////////////////////////////

(function(global, $data, undefined) {




  function registerEdmTypes() {

    function Edm_Boolean() { };
    $data.Container.registerType('Edm.Boolean', Edm_Boolean);
    $data.Container.mapType(Edm_Boolean, $data.Boolean);

    function Edm_Binary() { };
    $data.Container.registerType('Edm.Binary', Edm_Binary);
    $data.Container.mapType(Edm_Binary, $data.Blob);

    function Edm_DateTime() { };
    $data.Container.registerType('Edm.DateTime', Edm_DateTime);
    $data.Container.mapType(Edm_DateTime, $data.Date);

    function Edm_DateTimeOffset() { };
    $data.Container.registerType('Edm.DateTimeOffset', Edm_DateTimeOffset);
    $data.Container.mapType(Edm_DateTimeOffset, $data.Integer);

    function Edm_Time() { };
    $data.Container.registerType('Edm.Time', Edm_Time);
    $data.Container.mapType(Edm_Time, $data.Integer);

    function Edm_Decimal() { };
    $data.Container.registerType('Edm.Decimal', Edm_Decimal);
    $data.Container.mapType(Edm_Decimal, $data.Number);

    function Edm_Single() { };
    $data.Container.registerType('Edm.Single', Edm_Single);
    $data.Container.mapType(Edm_Single, $data.Number);

    function Edm_Double() { };
    $data.Container.registerType('Edm.Double', Edm_Double);
    $data.Container.mapType(Edm_Double, $data.Number);

    function Edm_Guid() { };
    $data.Container.registerType('Edm.Guid', Edm_Guid);
    $data.Container.mapType(Edm_Guid, $data.String);

    function Edm_Int16() { };
    $data.Container.registerType('Edm.Int16', Edm_Int16);
    $data.Container.mapType(Edm_Int16, $data.Integer);

    function Edm_Int32() { };
    $data.Container.registerType('Edm.Int32', Edm_Int32);
    $data.Container.mapType(Edm_Int32, $data.Integer);

    function Edm_Int64() { };
    $data.Container.registerType('Edm.Int64', Edm_Int64);
    $data.Container.mapType(Edm_Int64, $data.Integer);

    function Edm_Byte() { };
    $data.Container.registerType('Edm.Byte', Edm_Byte);
    $data.Container.mapType(Edm_Byte, $data.Integer);

    function Edm_String() { };
    $data.Container.registerType('Edm.String', Edm_String);
    $data.Container.mapType(Edm_String, $data.String);

  };
  registerEdmTypes();


  })(window, $data);
```

Not sure if I'm doing something wrong here or if that would is expected for the RC1 version.

Every feedback welcome,

Rainer

**Comment by PeterAronZentai** [comment]
I am on it :-) 

The first issue regarding missing attribs and one-side-only referenced navigation props seems harmless albeit annoying. I will however check if everything works as expected. Expect an answer within a day.

The second issue seems to be related to the old OData namespace that is xmlns="http://schemas.microsoft.com/ado/2007/05/edm" instead of xmlns="http://schemas.microsoft.com/ado/2008/09/edm" found in the specs. This recently has been reported from other sources so I modify the converter a bit to be less meticulous on the namespace versions as long as it can process it. Expect a fix within hours, one day top.

**Comment by PeterAronZentai** [comment]
Please, also note that we just released 1.0.4 with many fixes. JaySvcUtil is still affected though.

**Comment by RainerAtSpirit** [comment]
Noticed that :). I'm trying to integrate it with the Netflx service in a Metro app on Win8 right now.

Thanks for looking into it.

**Comment by PeterAronZentai** [comment]
Quickrep on Netflix service:
I could make it working in IE9 browser after commenting out 3 lines in the generated data context, where inverseProperty was incorrectly set empty. 

```
//'Discs': { type: 'Array', elementType: 'Netflix.Catalog.v2.Title', inverseProperty: '' },
//'Episodes': { type: 'Array', elementType: 'Netflix.Catalog.v2.Title', inverseProperty: '' },
//'Seasons': { type: 'Array', elementType: 'Netflix.Catalog.v2.Title', inverseProperty: '' }
```

Consequently these fields became unavailable by this but this way I could get the top 5 Titles and all Genres as expected. I assume to other sets and the functionimports to work poperly. This will be fixed together with subissue no2.

I could not however validate if all these work in a Metro style app. Everything should because of the HTML5 feature set but I haven't tried it myself. We would love some feedback on it. In the mean time we are into putting together a test line to windows 8/ metro apps so that they could be tested with the build. Any advice on that?

Here is my code:

<pre>

<html xmlns="http://www.w3.org/1999/xhtml">
<head>
    <title></title>
    <script type="text/javascript" src="Scripts/jquery-1.7.2.js"></script>
    <script type="text/javascript" src="Scripts/datajs-1.0.3.js"></script>
    <script type="text/javascript" src="Scripts/JayData.js"></script>
    <script type="text/javascript" src="Netflix.js"></script>
    <script>
        jQuery.support.cors = true;
        var context = new System.Data.Objects.NetflixCatalog({ name: 'oData', oDataServiceHost: 'http://odata.netflix.com/v2/catalog' });
        
        context.Titles.take(5).toArray(function (items) {
            alert(items);
        });

        context.Genres.toArray(function (items) {
            alert(items);
        });
    </script>
</head>
<body>

</body>
</html
</pre>

**Comment by RainerAtSpirit** [comment]
Thanks Peter,

I'm back on IIS7.5 at the momnet as I wanted to have your example up and running before moving on. Everything looks fine and I see the corect requests to netflix, but I'm getting a `Origin http://localhost:2146 not found in Access-Control-Allow-Origin header.` message in FF, that I couldn't track down yet. 
Anyway I doubt that this will be an issue with an metro app as I'm already talking to the netflix service.
Here's a small blog that talks about it http://codefoster.com/post/2012/04/11/Simple-Netflix-Browser-in-Windows-8-JavaScriptndash;Part-1.aspx

I've to admit that my metro apps won't go any further than this at the moment, but I thought it would be a good opportunity to throw JayData to it just to see how it goes :).

**Comment by PeterAronZentai** [comment]
Yes, FF does not support this,and IE only for trusted  or local sites. 

According to your post you'll need to include genres to titles. In case you did not already know this can be achieved by the .include(predicate/string) frame expression like this:

<pre>
 netflix.Titles.inlucde("Genres").take(100).toArray( function(titles) {... })
</pre>


My next concern is that JayData uses jQuery that ultimately uses window.XMLHttpRequest for handling network. Digging the web tells that albeit working the window.xmlhttprequest will have cross origin issues (at leas in january 2012), while WinJS.xhr will not. 
I've opened a separate issue #10 on this.

I understand that your purpose is to evaluate and write a publication, and I would love seeing JayData in action in your blog. So if you would not mind let us solve any upcoming issues so that the community could get a cool article about how to do it in  Metro / Windows 8. 

We are also passionate lovers of the Metro UI, perhaps you could notice in our JayScrum technology sample.

**Comment by RainerAtSpirit** [comment]
Just to clarify the blog is not mine, so my purpose is not writing publications ;-). I'm just following along what Jeremy Foster wrote up. With that said I agree that using WinJS.xhr would be the way to go for Metro style apps.
For Metro UI like HTML5 cross browser apps either JSONP or something along the line as described here http://www.kendoui.com/blogs/teamblog/posts/11-10-04/using_cors_with_all_modern_browsers.aspx should do the job. 

I like the Metro UI style that I've seen in the JayScrum preview video and would be interested leveraging that on top of an existing SharePoint OData example (http://www.spirit.de/demos/metro/ServiceFiles/DynMetroStyle.aspx). 
Not a blog I've to admit, but at least a Proof of concept ;-). Please ping me offline if that's something you're interested in.

Rainer

**Comment by PeterAronZentai** [comment]
Hello Rainer,

But of course it is a very interesting and very much asked for aspect. HTML5 + Metro + OData + Sharepoint, I guess this will be one of the next gen enterprise web application architectures. How could we participate?

Yeah ok, I contact you "offline" :)

Regarding JSONP:  as it is a protocol level thing at this point JSONP can be used with JayData if you tell the jQuery part to use it. I'll grab some info on that and publish in a post just for the sake of completeness. There is one problem with JSONP however: it will not support write operations as no support for methods other then GET.

**Comment by RainerAtSpirit** [comment]
Yeap, JSONP is not the answer if you need C(R)UD. For OData there's a nice summary of options at http://datajs.codeplex.com/wikipage?title=Cross%20Domain%20Requests&referringTitle=Documentation

**Comment by RainerAtSpirit** [comment]
Just as an update on the original thread ;-). I tried to process a SharePoint 2010 listdata.svc/$metadata and run into the same issue described in 1. After updating the namespace per your suggestion the file was processed, but with some errors/warnings.

https://gist.github.com/2853140

Might be good to use that as another test case for **1.0.5**

**Comment by PeterAronZentai** [comment]
The original issue with the namespace versions has been resolved with 1.0.5
Also the incorrect inverseProperty detection with empty values has been resolved with 1.0.5

I open a new issue for investigate with the Sharepoint metadata whether it is still problematic or not.

**Comment by RainerAtSpirit** [comment]
Thanks. Found your Netflix demo so I can confirm that this is working.
